### PR TITLE
net-misc/wide-dhcpv6: new package

### PR DESCRIPTION
**Pull request by DarthGandalf**
Heavily based on Thomas Schneider's version from qsx overlay.

Differences from the overlay:
* added OpenRC services for client and server
* import Debian patches directly, instead of listing them one by one
* make sla_len and sla_id optional in config (by including the newer
  Debian patches)

Package-Manager: Portage-3.0.18, Repoman-3.0.2
Signed-off-by: Alexey Sokolov <alexey+gentoo@asokolov.org>
Closes: https://bugs.gentoo.org/761532

### CODE REVIEW

File: net-misc/wide-dhcpv6/wide-dhcpv6-20080615_p23.ebuild [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+LICENSE="BSD"
+SLOT="0"
+KEYWORDS="~amd64"
+
+BDEPEND="
+	sys-devel/bison
+	sys-devel/flex
+"
+
+S="${WORKDIR}/wide-dhcpv6-${MY_PV}"
```
Comment (SoapGentoo) [inline net-misc/wide-dhcpv6/wide-dhcpv6-20080615_p23.ebuild:28]: put right after `SRC_URI`

File: net-misc/wide-dhcpv6/wide-dhcpv6-20080615_p23.ebuild [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+MY_PV="$(ver_cut 1)"
+MY_DEB_PV="$(ver_cut 1)-$(ver_cut 3)"
+
+DESCRIPTION="WIDE project DHCPv6 client and server"
+HOMEPAGE="https://wide-dhcpv6.sourceforge.net/"
+SRC_URI="
+	mirror://sourceforge/wide-dhcpv6/wide-dhcpv6-${MY_PV}.tar.gz
+	mirror://debian/pool/main/w/wide-dhcpv6/wide-dhcpv6_${MY_DEB_PV}.debian.tar.xz
+"
+LICENSE="BSD"
```
Comment (SoapGentoo) [inline net-misc/wide-dhcpv6/wide-dhcpv6-20080615_p23.ebuild:19]: empty line before `LICENSE`

File: net-misc/wide-dhcpv6/wide-dhcpv6-20080615_p23.ebuild [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	insinto /etc/wide-dhcpv6
+	doins dhcp6c.conf.sample dhcp6s.conf.sample
+	dodoc "${FILESDIR}/README.gentoo"
+	systemd_newunit "${FILESDIR}/dhcp6c-AT.service" "dhcp6c@.service"
+	systemd_newunit "${FILESDIR}/dhcp6s-AT.service" "dhcp6s@.service"
+	newinitd wide-dhcp6c.init wide-dhcp6c
+	newinitd wide-dhcp6s.init wide-dhcp6s
+	newconfd "${FILESDIR}/wide-dhcp6c.conf" wide-dhcp6c
+	newconfd "${FILESDIR}/wide-dhcp6s.conf" wide-dhcp6s
+	einstalldocs
```
Comment (SoapGentoo) [inline net-misc/wide-dhcpv6/wide-dhcpv6-20080615_p23.ebuild:65]: put some empty lines in this block, it's all mushed together

**Review by SoapGentoo — CHANGES_REQUESTED** [review]

**Comment by DarthGandalf** [comment]
Fixed

**Comment by SoapGentoo** [comment]
The last release of WIDE was in 2008, and Debian maintains it with a bunch of patches. Adding this as a new package is a real issue. @juippis @mgorny?

**Comment by DarthGandalf** [comment]
To clarify: this is the only DHCPv6 client which worked  well for me. All the other ones I tried had various issues (didn't like prefix length for DHCP PD, etc)
This is why I'd like to see it in repo.

File: net-misc/wide-dhcpv6/wide-dhcpv6-20080615_p23.ebuild [hunk 4]
```diff
@@ -0,0 +1,77 @@
+# Copyright 2021 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+# Copyright 2020 Thomas Schneider <qsx@chaotikum.eu>
```
Comment (SoapGentoo) [inline net-misc/wide-dhcpv6/wide-dhcpv6-20080615_p23.ebuild:4]: also, this won't fly

**Comment by juippis** [comment]
If someone sees any use for this, sure I don't mind. 
But seeing it's from 2008, and looking at these massive required patches:
https://sources.debian.org/src/wide-dhcpv6/20080615-23/debian/patches/
```

&nbsp; | - | rw-r--r-- | 4,802 | 0001-Fix-manpages.patch
-- | -- | -- | -- | --
&nbsp; | - | rw-r--r-- | 914 | 0002-Don-t-strip-binaries.patch
&nbsp; | - | rw-r--r-- | 1,549 | 0003-Close-inherited-file-descriptors.patch
&nbsp; | - | rw-r--r-- | 226,218 | 0004-GNU-libc6-fixes.patch
&nbsp; | - | rw-r--r-- | 982 | 0005-Update-ifid-on-interface-restart.patch
&nbsp; | - | rw-r--r-- | 16,932 | 0006-Add-new-feature-dhcp6c-profiles.patch
&nbsp; | - | rw-r--r-- | 4,254 | 0007-Adding-ifid-option-to-the-dhcp6c.conf-prefix-interfa.patch
&nbsp; | - | rw-r--r-- | 6,476 | 0008-Close-file-descriptors-on-exec.patch
&nbsp; | - | rw-r--r-- | 712 | 0009-Fix-renewal-of-IA-NA.patch
&nbsp; | - | rw-r--r-- | 1,413 | 0010-Call-client-script-after-interfaces-have-been-update.patch
&nbsp; | - | rw-r--r-- | 3,733 | 0011-resolv-warnings-so-as-to-make-blhc-and-gcc-both-happ.patch
&nbsp; | - | rw-r--r-- | 564 | 0012-fix-a-redefined-YYDEBUG-warning-of-gcc-for-the-code-.patch
&nbsp; | - | rw-r--r-- | 5,616 | 0013-added-several-comments-examples-by-Stefan-Sperling.patch
&nbsp; | - | rw-r--r-- | 11,555 | 0014-Support-to-build-on-kFreeBSD-n-GNU-Hurd-platform.patch
&nbsp; | - | rw-r--r-- | 1,724 | 0015-a-bit-info-to-logger-when-get-OPTION_RECONF_ACCEPT.patch
&nbsp; | - | rw-r--r-- | 894 | 0016-fix-typo-in-dhcp6c.8-manpage.patch
&nbsp; | - | rw-r--r-- | 1,390 | 0017-Remove-unused-linking-with-libfl.patch
&nbsp; | - | rw-r--r-- | 4,478 | 0018-dhcpv6-ignore-advertise-messages-with-none-of-reques.patch
&nbsp; | - | rw-r--r-- | 2,315 | 0019-Server-should-not-bind-control-port-if-there-is-no-s.patch
&nbsp; | - | rw-r--r-- | 3,940 | 0020-Adding-option-to-randomize-interface-id.patch
&nbsp; | - | rw-r--r-- | 1,117 | 0021-Make-sla-len-config-optional.patch
&nbsp; | - | rw-r--r-- | 2,027 | 0022-Make-sla-id-config-optional.patch

```
I am not personally convinced. This looks like it can break any minute. 
https://repology.org/projects/?search=wide-dhcp pretty much only available in distros shipping ancient software (Debian derivatives)

Kudos for the systemd and openrc files though.

**Comment by DarthGandalf** [comment]
> But seeing it's from 2008, and looking at these massive required patches:

That's why I consider debian the upstream now, and pulled their patches.

> distros shipping ancient software (Debian derivatives)

Debian stable is ancient, Debian sid is not that old. [Their changelog](https://metadata.ftp-master.debian.org/changelogs//main/w/wide-dhcpv6/wide-dhcpv6_20080615-23_changelog) says the latest update is from Sept 2020.

Also Fedora is not a derivative of Debian :P

**Comment by SoapGentoo** [comment]
```
dhcpcd-9.4.0 starting
dev: loaded udev
DUID 00:01:00:01:1b:ad:de:e0:f0:de:f1:d5:bd:bd
eth0: IAID 00:00:00:01
eth0: IA type 25 IAID 00:00:00:02
eth0: soliciting a DHCPv6 lease
dhcpcd_prestartinterface: Operation not possible due to RF-kill
wlan0: waiting for carrier
eth0: soliciting an IPv6 router
eth0: Router Advertisement from fe80::201:5cff:fe87:6d71
eth0: no global addresses for default route
eth0: adding route to 2001:4060:c002:3::/64 via fe80::201:5cff:fe87:6d71
eth0: ADV 2001:4060:c005:3e00::/56 from fe80::201:5cff:fe87:6d71
eth0: REPLY6 received from fe80::201:5cff:fe87:6d71
eth0: renew in 7200, rebind in 14400, expire in 86400 seconds
lo: adding reject route to 2001:4060:c005:3e00::/56
eth0: delegated prefix 2001:4060:c005:3e00::/56
```
clearly `dhcpcd-9.4.0` does PD fine, and it uses the whole `/56` prefix I get from my ISP, so I'm not convinced that this use case definitely requires adding a dead DHCPv6 server.

**Comment by DarthGandalf** [comment]
```
$ sudo dhcpcd -6 -T eth1
Password: 
dhcpcd-9.4.0 starting
DUID 00:01:00:01:20:03:d3:ae:b8:27:eb:84:bc:e5
eth1: waiting for carrier
eth1: carrier acquired
eth1: IAID c6:c3:a6:21
eth1: adding address fe80::3821:9310:42f6:37c4
eth1: soliciting an IPv6 router
ipv6nd_sendrsprobe: Operation not permitted
ipv6nd_sendrsprobe: Operation not permitted
ipv6nd_sendrsprobe: Operation not permitted
ipv6nd_sendrsprobe: Operation not permitted
eth1: no IPv6 Routers available
timed out
dhcpcd exited
```

I'm glad dhcpcd works for you. But it doesn't work for me with my ISP.

**Comment by DarthGandalf** [comment]
Thanks for helping me debug dhcpcd, it works for me now.

**Pull request closed without merging**